### PR TITLE
8313374: --enable-ccache's CCACHE_BASEDIR breaks builds

### PR DESCRIPTION
**Pull request by jankratochvil**
https://bugs.openjdk.org/browse/JDK-8313374
--enable-ccache's CCACHE_BASEDIR breaks builds

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313374](https://bugs.openjdk.org/browse/JDK-8313374): --enable-ccache's CCACHE_BASEDIR breaks builds (**Bug** - P4)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15080/head:pull/15080` \
`$ git checkout pull/15080`

Update a local copy of the PR: \
`$ git checkout pull/15080` \
`$ git pull https://git.openjdk.org/jdk.git pull/15080/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15080`

View PR using the GUI difftool: \
`$ git pr show -t 15080`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15080.diff">https://git.openjdk.org/jdk/pull/15080.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15080#issuecomment-1656738431)

### CODE REVIEW

**Comment by dholmes-ora** [comment]
@jankratochvil  can please update the JBS issue to explain how you went from the problem description there to the proposed fix. Thanks

**Comment by TheShermanTanker** [comment]
@jankratochvil The change here will force ccache to write absolute paths for all compilations, which may cause cached object file comparisons to fail sometimes, and I'm not sure we'd want that? Have you tried running make clean before recompilations? (Note that there are 2 different targets for this, make clean for cleaning build artifacts and make dist-clean which removes almost everything in the build directory).

Weird bugs can happen sometimes if clean is not run, as stated in the warning message each time configure is run. I also read the issue in the tracker, and I don't think removing --enable-ccache entirely is correct, as we need that option to check for certain things (like if ccache can handle precompiled headers) and also disable aliasing ccache as the compiler during the build for that exact reason. I'm also not too sure how removing CCACHE_BASEDIR helps fix the issue, since all it does it change the command passed to the real compiler, could you elaborate on that slightly more?

**Comment by TheShermanTanker** [comment]
Woah! Slow down a little, I wasn't asking you to remove the change entirely, just to elaborate on what the change does to achieve the fix, since it's a little unclear to me

**Comment by jankratochvil** [comment]
OK so the real problem is:
- OpenJDK makefiles operate with absolute paths to the target objects files.
- Dependency files (`*.d`) switch to relative paths to everything by turning on `CCACHE_BASEDIR`.
- `make` does not find a relative vs. absolute filename of the target to be the same as `make` does just a string comparison, not filesystem inode comparison.

When trying to fix it I have found `$(FILE_MACRO_CFLAGS)` is on my system `-fmacro-prefix-map=/home/user/jdk-src-dir/=`. I do not see why it was implemented this way by JDK-8226346 (and I cannot see this Bug). So I have removed it.

**Comment by jankratochvil** [comment]
Personally I am scared of `CCACHE_BASEDIR`, for example how debug info will look afterwards (I know there is some remapping but still). I was already troubleshooting/fixing multiple ccache bugs in the past so I find `CCACHE_BASEDIR` too bold. But if you want to keep it let's fix at least the known bugs.

**Comment by TheShermanTanker** [comment]
I see, I don't think we have to keep CCACHE_BASEDIR if it is not required, but we should test it to see how the build reacts to such a change, since it does change the behaviour of ccache which is a little bit nerve wracking. I unfortunately cannot see the bug either, perhaps either David or @erikj79 could help with why both options were implemented this way

File: make/autoconf/build-performance.m4 [hunk 1]
```diff
@@ -243,8 +243,12 @@ AC_DEFUN([BPERF_SETUP_CCACHE_USAGE],
       fi
     fi
 
+    if $ALLOW_ABSOLUTE_PATHS_IN_OUTPUT;then
```
Comment (dholmes-ora) [inline make/autoconf/build-performance.m4:246]: Does this even build? Surely you need `if test ...` here?

Comment (jankratochvil) [reply make/autoconf/build-performance.m4:246]: It works fine, I did test it.
```
make/autoconf/jdk-options.m4:    ALLOW_ABSOLUTE_PATHS_IN_OUTPUT="false"
make/autoconf/jdk-options.m4:    ALLOW_ABSOLUTE_PATHS_IN_OUTPUT="true"
make/autoconf/jdk-options.m4:    ALLOW_ABSOLUTE_PATHS_IN_OUTPUT="false"
make/autoconf/jdk-options.m4:    ALLOW_ABSOLUTE_PATHS_IN_OUTPUT="true"
```
I see other code is using the test:
```
make/autoconf/flags-cflags.m4:    if test "x$ALLOW_ABSOLUTE_PATHS_IN_OUTPUT" = "xfalse"; then
```
So I will change it.

**Comment by jankratochvil** [comment]
> I see, I don't think we have to keep CCACHE_BASEDIR if it is not required,

BTW `CCACHE_BASEDIR` is a caching improvement - without it builds in different directories never share the cache. Some existing `ccache` users may find the `CCACHE_BASEDIR` removal as a regresson. Although I doubt anyone is using `ccache` with OpenJDK as it has been a PITA without working dependencies.

**Comment by dholmes-ora** [comment]
@jankratochvil See the comments here: 

https://github.com/openjdk/jdk/blob/eaa42e853dbe45c5dd26f75d0b3c242ebebdd286/make/autoconf/flags-cflags.m4#L821

and before the change you made in NativeCompilation.gmk to understand why this is done.

**Comment by TheShermanTanker** [comment]
Ah, these changes were done for better reproducible builds, I had completely missed that. The relative paths mean that the compiled files aren't all using different absolute paths in their macros and so on, making the build more reproducible. Changing these would be rather problematic for a lot of things unfortunately

**Comment by dholmes-ora** [comment]
> Ah, these changes were done for better reproducible builds

Actually this pre-dates that but yes relative paths do support reproducible builds. :)

File: make/autoconf/build-performance.m4 (hunk 1 above)
Comment (dholmes-ora) [reply make/autoconf/build-performance.m4]: Yes please always use the existing style in the build files - we know that they work correctly everywhere, whereas other approaches may not.

**Review by dholmes-ora — COMMENTED** [review]
That certainly looks much clearer in terms of what is being done and why - thanks. Now just need someone with ccache experience to validate the actual fix.

**Comment by jankratochvil** [comment]
I doubt there is anyone as it did not work. :-)

**Review by TheShermanTanker — CHANGES_REQUESTED** [review]

File: make/common/NativeCompilation.gmk [hunk 2]
```diff
@@ -216,6 +216,19 @@ DEPENDENCY_TARGET_SED_PATTERN := \
...
+# CCACHE_BASEDIR needs fix-deps-file as makefiles use absolute filenames for
+# object files while CCACHE_BASEDIR will make ccache relativize all paths for
+# its compiler. The compiler then produces relative dependency files.
+# make does not know a relative and absolute filename is the same so it will
+# ignore such dependencies.
+ifneq ($(CCACHE), )
+  NEED_FIX_DEPS_FILE = 1
+endif
+
+ifdef NEED_FIX_DEPS_FILE
```
Comment (TheShermanTanker) [inline make/common/NativeCompilation.gmk:231]:
```suggestion
ifeq ($(call Or, $(if $(CCACHE), true), $(call equals $(ALLOW_ABSOLUTE_PATHS_IN_OUTPUT), false))-$(FILE_MACRO_CFLAGS), true-)
  # CCACHE_BASEDIR needs fix-deps-file as makefiles use absolute filenames for
  # object files while CCACHE_BASEDIR will make ccache relativize all paths for
  # its compiler. The compiler then produces relative dependency files.
  # make does not know a relative and absolute filename is the same so it will
  # ignore such dependencies.
```

**Comment by TheShermanTanker** [comment]
I've left a code suggestion that hopefully fixes the issue of it not working. You can move the comment about ccache anywhere you want otherwise

File: make/common/NativeCompilation.gmk [hunk 3]
```diff
@@ -215,7 +215,7 @@ DEPENDENCY_TARGET_SED_PATTERN := \
 # The fix-deps-file macro is used to adjust the contents of the generated make
 # dependency files to contain paths compatible with make.
 #
-ifeq ($(ALLOW_ABSOLUTE_PATHS_IN_OUTPUT)-$(FILE_MACRO_CFLAGS), false-)
+ifeq ($(ALLOW_ABSOLUTE_PATHS_IN_OUTPUT), false)
```
[the remark this answers is not in the document]
Comment (jankratochvil) [reply make/common/NativeCompilation.gmk]: This part is no longer in the patch.

**Comment by jankratochvil** [comment]
@TheShermanTanker made a suggestion:
```
ifeq ($(call Or, $(if $(CCACHE), true), $(call equals $(ALLOW_ABSOLUTE_PATHS_IN_OUTPUT), false))-$(FILE_MACRO_CFLAGS), true-)
```
I got it by mail but I do not see it here - has it been deleted? I admit I prefer how it is now, this oneliner looks a bit cryptic to me.

**Comment by TheShermanTanker** [comment]
Yes, it's been deleted since I deemed it too verbose for this use case

**Comment by erikj79** [comment]
I understand why you need `fix-deps-file` when using `ccache`, but do you also need `MakeCommandRelative`?

I'm not surprised that ccache support has bit rotted over the years as we aren't seeing much benefit from it in practice, so it's probably not used much. In the ideal case, it certainly speeds up the build a lot, but that case is very rare, at least in our build scenarios. I wouldn't mind removing it at this point, but if we can fix it with a simple patch like this, then that works too. Note that trying to setup ccache correctly without support in the build is quite tricky to get right. That's why I thought it necessary to handle it explicitly in the makefiles.

**Comment by jankratochvil** [comment]
> I understand why you need `fix-deps-file` when using `ccache`, but do you also need `MakeCommandRelative`?

I did try it without `MakeCommandRelative`. But it did not work. As the relative paths created by `ccache` itself will look somehow like `dir1/dir2/../../dir3/dir4` which again does not match the intended `dir3/dir4`.

> I'm not surprised that ccache support has bit rotted over the years as we aren't seeing much benefit from it in practice,

I do see a big benefit from it. When changing branches by `git checkout` in a single directory it gets perfectly cached.

> Note that trying to setup ccache correctly without support in the build is quite tricky to get right. That's why I thought it necessary to handle it explicitly in the makefiles.

It is not tricky, I am happily using `ccache` for many projects including LLVM, GCC, GDB and others. You only do not have it cached across different directories (without `CCACHE_BASEDIR`) but as I have only 2-3 build trees for each projects (my mind can handle working only on 2-3 different branches at once) it does not matter much.

File: make/common/NativeCompilation.gmk [hunk 4]
```diff
@@ -216,6 +216,19 @@ DEPENDENCY_TARGET_SED_PATTERN := \
...
+  NEED_FIX_DEPS_FILE = 1
+endif
+
+# CCACHE_BASEDIR needs fix-deps-file as makefiles use absolute filenames for
+# object files while CCACHE_BASEDIR will make ccache relativize all paths for
+# its compiler. The compiler then produces relative dependency files.
+# make does not know a relative and absolute filename is the same so it will
+# ignore such dependencies.
+ifneq ($(CCACHE), )
+  NEED_FIX_DEPS_FILE = 1
```
Comment (erikj79) [inline make/common/NativeCompilation.gmk:228]: If the ccache case needs both of these macros, then the variable name should reflect that better. Maybe `REWRITE_PATHS_RELATIVE`?

**Comment by kimbarrett** [comment]
> I doubt there is anyone as it did not work. :-)

Um, I've been using it for years without noticing any problems, and measured a frequent speedup for my normal use.

File: make/common/NativeCompilation.gmk [hunk 5]
```diff
@@ -216,6 +216,19 @@ DEPENDENCY_TARGET_SED_PATTERN := \
...
+# CCACHE_BASEDIR needs fix-deps-file as makefiles use absolute filenames for
+# object files while CCACHE_BASEDIR will make ccache relativize all paths for
+# its compiler. The compiler then produces relative dependency files.
+# make does not know a relative and absolute filename is the same so it will
+# ignore such dependencies.
+ifneq ($(CCACHE), )
+  REWRITE_PATHS_RELATIVE = 1
+endif
+
+ifdef REWRITE_PATHS_RELATIVE
```
Comment (erikj79) [inline make/common/NativeCompilation.gmk:231]: We generally try to protect the makefiles from misbehaving due to environment variables being set with unexpected values. We also prefer `true`/`false` for boolean values. In this case it would be better to initialize REWRITE_PATHS_RELATIVE to `false`, set it to `true` in the conditionals and check equals here.
```
ifeq ($(REWRITE_PATHS_RELATIVE), true)
```

**Review by erikj79 — APPROVED** [review]

**Comment by kimbarrett** [comment]
I don't understand what problem this PR is fixing.  I see assertions from the
PR author that using ccache is broken with OpenJDK, or it breaks dependencies,
or something.  Yet I've been using ccache for years for my local development
builds, and have not encountered any problems, only build time improvements.

**Comment by jankratochvil** [comment]
> I don't understand what problem this PR is fixing.

Does the reproducer work for you? https://bugs.openjdk.org/browse/JDK-8313374

**Comment by kimbarrett** [comment]
> > I don't understand what problem this PR is fixing.
> 
> Does the reproducer work for you? https://bugs.openjdk.org/browse/JDK-8313374

Yes, it works for me. Or at least, with the commands and setup I normally use,
it works.  One difference is that I do out-of-tree builds, so I retried with
an in-tree build, but that still worked.  Another difference is that I'm using
an in-house (Oracle) build wrapper called jib, which does things like ensuring
the compiler and other tools we support are being used.  I don't even remember
how to do a build without it.

**Comment by jankratochvil** [comment]
Maybe your setup does not reproduce this problem. Still the most simple ccache setup does reproduce the problem. And the problem is known and described above. So why it should not be fixed?
I sure know how to workaround this bug. As this fix has not yet landed I use `--disable-ccache` and hide from the `configure` script I am in fact using ccache.
```
#! /bin/bash
# openjdk: export PATH="$(echo "$PATH"|sed 's#:/usr/lib64/ccache:#:'$HOME'/ccache:#')";bash configure --disable-precompiled-headers --disable-ccache
set -ex
rm -rf ~/ccache
mkdir ~/ccache
cd ~/ccache
for i in /usr/lib64/ccache/*;do
  j=`basename $i`
  echo -e '#! /bin/bash\nexport PATH="$(echo "$PATH"|sed s#:$HOME/ccache:#:#)"\nexec '$i' "$@"' >$j
  chmod +x $j
done
echo done
```

**Comment by erikj79** [comment]
I tried this myself today and here are my findings. If the output dir is a subdir of the `CCACHE_BASEDIR` the issue reproduces. The source/header files in the *.d files can be relative without impacting how make resolves them, but if the object files are relative, then make doesn't understand that they should match the absolute files used in the rest of the makefiles.

Kim is likely not seeing this because in Oracle builds, because we add a custom repository outside of the OpenJDK repository, we typically put the build dir outside of the OpenJDK repo, and so outside of `CCACHE_BASEDIR`. This is actually a bug with the ccache configuration. We should change the definition of `CCACHE_BASEDIR` to be `$WORKSPACE_ROOT` instead of `$TOPDIR` (otherwise source files in the Oracle repo will not be treated the same as source files in the OpenJDK repo).

So in summary, this fix is needed, and we have a different but with ccache handling for Oracle builds that hid it from us.
